File: 2024/Day 2 - Red-Nosed Reports/Day2.cpp
#include<iostream>
#include<fstream>
#include<sstream>
using namespace std;

int Analyzer(const vector<int> &report)
{
    int good = 0;
    bool safe = false;
    for(int i=0; i<report.size()-1; i++)
    {
        if(report[0]>report[1] && report[i]>report[i+1] && report[i]-report[i+1]<=3)
            safe = true;
        else if (report[0]<report[1] && report[i]<report[i+1] && report[i+1]-report[i]<=3)
            safe = true;
        else
        {
            safe = false;
            break;
        }
    }
    if(safe == true)
        good = 1;
    return good;
}

int SecondAnalyzer(const vector<int> &report)
{
    int secondgood = 0;
    if(!(Analyzer(report) == 1))
    {
        for(int i=0; i<report.size(); i++)
        {
            vector<int> newline = report;
            newline.erase(newline.begin()+i);
            if(Analyzer(newline) == 1)
            {
                secondgood = 1;
                break;
            }
        }
    }
    return secondgood;
}

int main()
{
    ifstream datafile("data2.txt");
    string line;
    vector<vector<int>> fullreport;
    int result = 0;
    int result2 = 0;
    
    while(getline(datafile, line))
    {
        stringstream s(line);
        vector<int> report;
        int level;
        while(s>>level)
        {
            report.push_back(level);
        }
        fullreport.push_back(report);
    }
    
    for(auto report:fullreport)
    {
        result += Analyzer(report);
        result2 += SecondAnalyzer(report);
    }
    cout<< result <<endl;
    cout<< result + result2 <<endl;
}

File: 2024/Day 2 - Red-Nosed Reports/Day2.py
with open('data2.txt') as file:
    data = file.readlines()

def Analyzer(line):
    good = 0
    safe = False
    for i in range(len(line)-1):
        if (int(line[0])>int(line[1])) and (int(line[i])>int(line[i+1])) and (int(line[i])-int(line[i+1])<=3)== True:
            safe = True
        elif (int(line[0])<int(line[1])) and (int(line[i])<int(line[i+1])) and (int(line[i+1])-int(line[i])<=3)== True:
            safe = True
        else:
            safe = False
            break
    
    if safe==True:
        good = 1
    return good

def SecondAnalyzer(line):
    secondgood = 0
    if not Analyzer(line) == 1:
        for i in range(len(line)):
            newline = line.copy()
            newline.pop(i)
            if Analyzer(newline) == 1:
                secondgood = 1
                break
    return secondgood
    
report = 0
report2 = 0
for line in data:
    line = line.strip().split()
    report += Analyzer(line)
    report2 += SecondAnalyzer(line)
print(report)
print(report+report2)            
